### PR TITLE
Autogenerate explorer links in address lists

### PR DESCRIPTION
**Pull request by fedgiac**
# Description

Reviewing the addition or removal of new chains in the contract addresses is always difficult to review. See for example [this PR](https://github.com/cowprotocol/docs/issues/538): there are lines with a thousand characters getting modified and it's very easy to miss differences that could [lead to issues](https://github.com/cowprotocol/docs/pull/543).

I tried two ways to work around this. One is better table formatting, but the solution I found (idea from [here](https://github.com/facebook/docusaurus/discussions/5155#discussioncomment-8036773)) requires relying on some relatively obscure package [remark-grid-tables](https://github.com/zestedesavoir/zmarkdown/tree/master/packages/remark-grid-tables). I didn't want to add an unknown dependency.

The other way was by relying on the fact that Docusaurus uses React. This PR is my attempt to use React to make adding chains easier.

This PR is my proposal on how to tackle this issue. I've never done any react before so the code is likely low quality and I don't fully understand what I'm doing. Please review carefully and question every choice!

Ideas for the future: it would be great to add the chain icon next to the name, so it's easier to find the chain you need.

# Changes

- Added new function to build links to explorer for a single contract. It has some minimal customization via option parameters.
- Used new function in every place I could. (There are still a lot of "raw" Etherscan link around in .md files. I think that's ok, if we feel the need we can migrate them to use this function.)

# Known issues

Lens explorer doesn't follow the same anchor format used by Etherscan, unlike all other explorers.
This means that things like `https://explorer.lens.xyz/address/0x9008D19f58AAbD9eD0D60971565AA8510560ab41#code` doesn't link to the code, the anchor should be `#contract`.
I could add some code to handle this edge case, but for now I'd leave it as it is since it's just Lens. The link still works, it's just that it doesn't point to the code.

# How to test

Visually compare the preview of this PR with the current version. Click on a few links to confirm they point to the expected URL.

List:
- Periphery: [current](https://docs.cow.fi/cow-protocol/reference/contracts/periphery), [new](https://docs-git-autogenerate-explorer-links-cowswap.vercel.app/cow-protocol/reference/contracts/periphery).
- Core: [current](https://docs.cow.fi/cow-protocol/reference/contracts/core), [new](https://docs-git-autogenerate-explorer-links-cowswap.vercel.app/cow-protocol/reference/contracts/core).
- Signing schemes: [current](https://docs.cow.fi/cow-protocol/reference/core/signing-schemes), [new](https://docs-git-autogenerate-explorer-links-cowswap.vercel.app/cow-protocol/reference/core/signing-schemes).

### CODE REVIEW

File: src/components/contract-addresses/index.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * explorer for the given address.
+ */
+export function explorerLinks(chains: string[], address: string) {
+  const componentsByChain = chains.map(
+    (chain, index) => {
+      const link = <Link to={explorerUrl(chain, address)}>{chain}</Link>;
+      return (index == 0) ? [link] : [", ", link];
+    }  
+  );
+  return componentsByChain.flat();
```
Comment (alfetopito) [inline src/components/contract-addresses/index.tsx]: Nitpick: use a reduce. Disclaimer, untested code.


```suggestion
  return chains.reduce(
    (acc, chain) => {
      if (acc.length) {
        acc.push(',')
      }
      acc.push(<Link to={explorerUrl(chain, address)}>{chain}</Link>)
      return acc
    }  
  , []);
```

Comment (fedgiac) [reply src/components/contract-addresses/index.tsx]: Applied in 4b0e81c5fcf0930322450bd74cb4f3fda3b40deb and tested. :white_check_mark:

File: src/components/contract-addresses/index.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+   * Defaults to the string ", ".
+   */
+  separator?: string | Component | undefined;
+  /**
+   * Which parameters to append after the URL. For example, #code points to the
+   * contract code.
+   * Defaults to #code.
+   */
+  urlTrailing?: string | undefined,
+}
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply src/components/contract-addresses/index.tsx:44]: Updated in b7765e1aa6c7b06fc2821fdbbbc41819ce664db6. Dropping `string` since that seems to be a valid `ReactNode` but keeping undefined for keeping the typing strictly explicit.

File: src/components/contract-addresses/index.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  return chains.reduce(
+    (acc, chain) => {
+      if (acc.length) {
+        acc.push(separator)
+      }
+      acc.push(<Link to={explorerUrl(chain, address, {urlTrailing})}>{chain}</Link>)
+      return acc
+    }
+  , []);
+}
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply src/components/contract-addresses/index.tsx:63]: Updated in b7765e1aa6c7b06fc2821fdbbbc41819ce664db6. Only using `React.ReactNode` since `React.ReactElement | string` are already captured in `React.ReactNode`.

**Review by anxolin — APPROVED** [review]
Way cleaner. 

I won't give you a hard time with this, but ideally we would reuse our SDK logic, since it has already modeled the networks and you don't need to repeat here.

https://github.com/cowprotocol/cow-sdk/tree/main/packages/config/src/chains/details

Also, ideally CoW Swap should move part of the logic to generate the explorer link to the SDK so it can be used here and anywhere else is needed.
https://github.com/cowprotocol/cowswap/blob/7016731bff3ddac7603803623cdb258563e10ca1/apps/explorer/src/components/common/BlockExplorerLink/BlockExplorerLink.tsx#L52

**Comment by fedgiac** [comment]
> ideally we would reuse our SDK logic, since it has already modeled the networks and you don't need to repeat here

I toyed a bit with this: in practice, this means replacing the keys of `explorerByChain` by a combination of `SupportedChainId` and `AdditionalTargetChainId`, and then replacing `{explorerLinks(["Ethereum", "Optimism"], "...")}` with `import {SupportedChainId, AdditionalTargetChainId} from 'sdk'; [...] {explorerLinks([SupportedChainId.MAINNET, AdditionalTargetChainId.OPTIMISM], "...")}`.

I think this makes it even less intuitive to edit the docs by non devs, it requires people to understand details of the SDK (why distinguishing `Supported` and `AdditionalTarget`, how do I find $random_chain, how do I add a chain that isn't present, is the sdk up to date). Considering how simple this code is, I'd say we can include the sdk in the future if we find additional use cases. (The nicest I can think of is that network information in the sdk already include the (token) logo of the network, which would be very nice to include here).

**Review by pretf00d — APPROVED** [review]
